File: src/main/java/com/busted_moments/client/features/ClickToCongratulateFeature.java
package com.busted_moments.client.features;

import com.busted_moments.client.util.PlayerUtil;
import com.busted_moments.core.Default;
import com.busted_moments.core.Feature;
import com.busted_moments.core.State;
import com.busted_moments.core.text.TextBuilder;
import com.wynntils.core.text.PartStyle;
import com.wynntils.core.text.StyledText;
import com.wynntils.handlers.chat.event.ChatMessageReceivedEvent;
import net.minecraft.network.chat.ClickEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static net.minecraft.ChatFormatting.AQUA;
import static net.minecraft.ChatFormatting.UNDERLINE;

@Default(State.ENABLED)
@Feature.Definition(name = "Click To Congratulate", description = "Click to send a congratulations message!")
public class ClickToCongratulateFeature extends Feature {
   private static final Pattern REGULAR_PATTERN = Pattern.compile("^\\[!] Congratulations to (?<player>.+) for reaching (?<type>.+) level (?<level>.+)!");
   private static final Pattern PROF_PATTERN = Pattern.compile("\\[!] Congratulations to (?<player>.+) for reaching level (?<level>.+) in (. )?(?<type>.+)!");

   @Value("Congratulations Message")
   @Tooltip("What message to use to congratulate people")
   private static String message = "Congratulations!";

   @SubscribeEvent
   public void onChatMessageReceivedEvent(ChatMessageReceivedEvent event) {
      StyledText text = event.getOriginalStyledText();

      Matcher matcher = text.getMatcher(REGULAR_PATTERN, PartStyle.StyleType.NONE);
      if (!matcher.matches() && !(matcher = text.getMatcher(PROF_PATTERN, PartStyle.StyleType.NONE)).matches())
         return;

      String player = matcher.group("player");
      if (!PlayerUtil.isPlayer(player)) return;

      event.setMessage(TextBuilder.of(event.getStyledText())
              .line()
              .append("Click to congratulate ", AQUA, UNDERLINE)
              .append(player, AQUA, UNDERLINE).append("!", AQUA, UNDERLINE)
              .onClick(
                      ClickEvent.Action.RUN_COMMAND,
                      "/msg %s %s".formatted(player, message)
              ).toComponent());
   }
}

File: src/main/kotlin/com/busted_moments/client/features/ClickToCongratulateFeature.kt
package com.busted_moments.client.features

import com.busted_moments.buster.api.Profile
import com.busted_moments.client.Patterns
import com.busted_moments.client.events.chat.originalMessage
import com.busted_moments.client.framework.config.annotations.Tooltip
import com.busted_moments.client.framework.config.entries.value.Value
import com.busted_moments.client.framework.events.Subscribe
import com.busted_moments.client.framework.features.Feature
import com.busted_moments.client.framework.text.Text
import com.busted_moments.client.framework.text.Text.matches
import com.busted_moments.client.framework.text.getValue
import com.wynntils.handlers.chat.event.ChatMessageEvent
import net.minecraft.network.chat.ClickEvent

object ClickToCongratulateFeature : Feature() {
    @Value("Message")
    @Tooltip(["The message to send when congratulating people!"])
    private var message: String = "Congratulations!"

    @Subscribe
    private fun ChatMessageEvent.Edit.on() {
        originalMessage.matches {
            any(
                Patterns.LEVEL_UP,
                Patterns.PROF_LEVEL_UP
            ) {
                val player by this
                if (!Profile.isValid(player ?: return))
                    return

                this@on.message = Text {
                    +it
                    newLine()

                    +"Click to congratulate ".aqua.underline.onClick(ClickEvent.Action.RUN_COMMAND, "/msg $player ${this@ClickToCongratulateFeature.message}")
                    +player!!
                    +"!"
                }
            }
        }
    }
}
